plugin/src/main/java/eu/maveniverse/maven/njord/plugin3/CheckArtifactsAvailabilityMojo.java

```diff
@@ -0,0 +1,255 @@
+                        "Invalid alt deployment repository syntax (supported is id::url): " + remoteRepository);
+            }
+            String id = split[0];
+            String url = split[1];
+            if (mode == RepositoryMode.SNAPSHOT) {
+                result = Optional.of(new RemoteRepository.Builder(id, "default", url)
+                        .setReleasePolicy(new RepositoryPolicy(false, null, null))
+                        .build());
+            } else {
+                result = Optional.of(new RemoteRepository.Builder(id, "default", url)
+                        .setSnapshotPolicy(new RepositoryPolicy(false, null, null))
+                        .build());
+            }
+        }
+        if (result.isPresent()) {
+            result = Optional.of(repositorySystem.newDeploymentRepository(
+                    mavenSession.getRepositorySession(), result.orElseThrow(J8Utils.OET)));
+        }
+        return result;
+    }
+
+    /**
+     * Uses Resolver to perform "existence checks" for given artifacts in given remote repository target. Based on
+     * parameters it may do "one pass" or "wait/poll".
+     */
+    protected void checkAvailability(Collection<Artifact> artifacts, RemoteRepository target)
+            throws IOException, MojoFailureException {
+        Duration waitTimeout = Duration.parse(this.waitTimeout);
+        Duration waitSleep = Duration.parse(this.waitSleep);
+        Map<Artifact, Boolean> artifactsMap = artifacts.stream().collect(Collectors.toMap(a -> a, a -> false));
+
+        Instant waitingUntil = Instant.now().plus(waitTimeout);
+        AtomicInteger toCheck = new AtomicInteger(artifactsMap.size());
+        if (wait) {
+            logger.info(
+                    "Waiting for {} artifacts to become available from {} (poll {}; timeout {})",
+                    artifactsMap.size(),
+                    target.getUrl(),
+                    waitSleep,
+                    waitTimeout);
+        } else {
+            logger.info("Checking for {} artifacts availability from {}", artifactsMap.size(), target.getUrl());
+        }
+        try (RepositoryConnector repositoryConnector =
+                repositoryConnectorProvider.newRepositoryConnector(mavenSession.getRepositorySession(), target)) {
+            while (toCheck.get() > 0) {
+                logger.info("Checking availability of {} artifacts (out of {}).", toCheck.get(), artifactsMap.size());
+                List<ArtifactDownload> artifactDownloads = new ArrayList<>();
+                artifactsMap.forEach((key, value) -> {
+                    if (!value) {
+                        ArtifactDownload artifactDownload = new ArtifactDownload(key, "njord", null, null);
+                        artifactDownload.setRepositories(Collections.singletonList(target));
+                        artifactDownload.setExistenceCheck(true);
+                        artifactDownloads.add(artifactDownload);
```

Comment:
While I agree for small projects, your approach is nice and give all the details you need, I think for large projects, we should break the loop as soon as we don't get one available instead of always testing all of them.

Reply:
This is a "batched request" (that does HEADs only!), and _very same mechanism_ like for resolving, it happens on multiple threads (see https://maven.apache.org/resolver/configuration.html `aether.connector,.basic.*` properties).

What happens here, is that we send a batch, that connector checks (on multiple threads), and then the "set" of artifacts is shrinking, as those found to exist are _removed_. And so on...

So like in [this gist](https://gist.github.com/cstamas/87a2db25076cdc48762c144a8c40378a), mojo will once check for all, then next check is only with missing, ideally the number of checks is dropping as more and more is published.

Reply:
Yeah, what I'm worried a bit about is sending thousands of requests every minute when the publication is delayed - which can happen, I have seen one that actually started syncing after 40 minutes.

But maybe it's all a matter of checking only after 30 minutes and checking every 5 minutes in our case. That wouldn't be too bad and would be good enough for us.